packages/apollo-server-fastify/src/fastifyApollo.ts

```diff
@@ -0,0 +1,92 @@
+import {
+  GraphQLOptions,
+  runHttpQuery,
+  convertNodeHttpToRequest,
+} from 'apollo-server-core';
+import { FastifyInstance, FastifyReply } from 'fastify';
+import { OutgoingMessage } from 'http';
+
+interface FastifyPluginOptions {
+  url?: string;
+  graphqlOptions: GraphQLOptions;
+}
+
+export function graphqlFastify(
+  app: FastifyInstance,
+  options: FastifyPluginOptions,
+  next: (err?: Error) => void,
+) {
+  if (!options || !options.graphqlOptions) {
+    throw new Error('Apollo Server requires options.');
+  }
+
+  if (arguments.length !== 3) {
+    throw new Error(
+      `Apollo Server expects exactly 3 argument, got ${arguments.length}`,
+    );
+  }
+
+  async function handler(request: any, reply: FastifyReply<OutgoingMessage>) {
+    const { method } = request.raw;
+    try {
+      const { graphqlResponse } = await runHttpQuery([request, reply], {
+        method: method,
+        options: options.graphqlOptions,
+        query: method === 'POST' ? request.body : request.query,
+        request: convertNodeHttpToRequest(request.raw),
+      });
+
+      reply
+        .type('application/json')
+        .code(200)
+        .header(
+          'Content-Length',
+          Buffer.byteLength(JSON.stringify(graphqlResponse), 'utf8'),
+        )
+        .send(JSON.parse(graphqlResponse));
```

Comment:
This is really not clear. Why are you serializing the response back-and-forth again? You are probably better off as not parsing/stringifying both the body and the request/response. This likely halves the throughput of this solution.

Reply:
Removing the serialization breaks most of the tests. I am assuming that the response need to be serialized.

Reply:
I mean that you are getting JSON via `runHttpQuery`, which I imagine it is parsed and stringified again, or maybe I'm mistaken by the API.